text/0000-nested-method-calls.md

```diff
@@ -0,0 +1,824 @@
+would compile, because (1) the method is `&mut self` and (2) the
+receiver is a simple variable reference `a`. Interestingly, now that
+we changed the method to `&mut self`, we can suddenly see the
+side-effects of evaluating the argument.
+
+On balance, it seems better to this author to have the borrow checker
+analysis be more complex than the desugaring and execution order.
+
+### Permit more things during the "restricted" period
+
+The current notion of a 'restricted' borrow is identical to a shared
+borrow. However, we could in principle permit *more* things during the
+restricted period -- basically we could permit anything that does not
+invalidate the reference we created. In that case, we might fruitfully
+enable two-phased borrows for shared references as well. In practice,
+this means that we could permit writes to the borrowed content (which
+are forbidden by this proposal). An example of code that would work as
+a result is the following:
+
+```rust
+// pretend you could define an inherent method on integers
+// for a second, just to keep code snippet simple
+impl i32 {
+    fn increment(&mut self, v: i32) -> i32 {
+        *self += v;
+        *self // returns new value
+    }
+}
+                                            
+fn foo() {
+    let mut x = 0;
+    let y = x.increment(x.increment(1)); // what result do you expect from this?
+    println!("{}", y);
+}
+```
+
+The call to `x.increment(x.increment(1))` would thus desugar to the following MIR:
+
+```
+tmp0 = &mut2 x;
+tmp1 = &mut2 x;
+tmp2 = 1;
+tmp3 = i32::increment(tmp1, tmp2); // activates tmp1
+i32::increment(tmp0, tmp3); // activates tmp0
+```
+
+Under the existing proposal, this is illegal, because `x` is
+considered "reserved" when `tmp1` is created, and an `&mut2` borrow is
+not permitted when the lvalue being borrowed has been reserved. If we
+made restrictions more permissive, we might accept this code; it would
+output `2`.
+
+We opted against this variation for several reasons:
+
+- It makes the borrow checker more complex by introducing not only
+  two-phase borrows, but a new set of restrictions that must be worked
+  out in detail. The current RFC leverages the existing category of
+  shared borrows.
+- The main gain here is the ability to intersperse two mutable calls
+  (as in the example), or to have an outer shared borrow with an inner
+  mutable borrow. In general, this implies that there is some careful
+  ordering of mutation going on here: in particular, the outer method
+  call will observe the state changes made by the inner calls. This
+  feels like a case where it is *helpful* to have the user pull the
+  two calls apart, so that their relative side-effects are clearly
+  visible.
+  
+Of course, it would be possible to loosen the rules in the future.
+  
+### A broader user of two-phase borrows
+
+The initial proposal for two-phased borrows (made in
+[this blog post][]) was more expansive. In particular, it aimed to
```

Comment:
Probably [this one](http://smallcultfollowing.com/babysteps/blog/2017/03/01/nested-method-calls-via-two-phase-borrowing/).